datafusion/core/src/physical_plan/file_format/row_filter.rs

```diff
@@ -0,0 +1,400 @@
+pub(crate) struct FilterCandidate {
+    expr: Expr,
+    required_bytes: usize,
+    can_use_index: bool,
+    projection: Vec<usize>,
+}
+
+/// Helper to build a `FilterCandidate`. This will do several things
+/// 1. Determine the columns required to evaluate the expression
+/// 2. Calculate data required to estimate the cost of evaluating the filter
+/// 3. Rewrite column expressions in the predicate which reference columns not in the particular file schema.
+///    This is relevant in the case where we have determined the table schema by merging all individual file schemas
+///    and any given file may or may not contain all columns in the merged schema. If a particular column is not present
+///    we replace the column expression with a literal expression that produces a null value.
+struct FilterCandidateBuilder<'a> {
+    expr: Expr,
+    file_schema: &'a Schema,
+    table_schema: &'a Schema,
+    required_column_indices: Vec<usize>,
+    non_primitive_columns: bool,
+    projected_columns: bool,
+}
+
+impl<'a> FilterCandidateBuilder<'a> {
+    pub fn new(expr: Expr, file_schema: &'a Schema, table_schema: &'a Schema) -> Self {
+        Self {
+            expr,
+            file_schema,
+            table_schema,
+            required_column_indices: vec![],
+            non_primitive_columns: false,
+            projected_columns: false,
+        }
+    }
+
+    pub fn build(
+        mut self,
+        metadata: &ParquetMetaData,
+    ) -> Result<Option<FilterCandidate>> {
+        let expr = self.expr.clone();
+        let expr = expr.rewrite(&mut self)?;
+
+        if self.non_primitive_columns || self.projected_columns {
+            Ok(None)
+        } else {
+            let required_bytes =
+                size_of_columns(&self.required_column_indices, metadata)?;
+            let can_use_index = columns_sorted(&self.required_column_indices, metadata)?;
+
+            Ok(Some(FilterCandidate {
+                expr,
+                required_bytes,
+                can_use_index,
+                projection: self.required_column_indices,
+            }))
+        }
+    }
+}
+
+impl<'a> ExprRewriter for FilterCandidateBuilder<'a> {
+    fn pre_visit(&mut self, expr: &Expr) -> Result<RewriteRecursion> {
+        if let Expr::Column(column) = expr {
+            if let Ok(idx) = self.file_schema.index_of(&column.name) {
+                self.required_column_indices.push(idx);
+
+                if !is_primitive_field(self.file_schema.field(idx)) {
+                    self.non_primitive_columns = true;
+                }
+            } else if self.table_schema.index_of(&column.name).is_err() {
+                // If the column does not exist in the (un-projected) table schema then
+                // it must be a projected column.
+                self.projected_columns = true;
+            }
+        }
+        Ok(RewriteRecursion::Continue)
+    }
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
```

Comment:
@alamb @thinkharderdev 
I find a issue caused by this rewrite in my work https://github.com/apache/arrow-datafusion/pull/3396
if we have two parquet, one has c1,c2, the other has c1,c3.
If the filter is c2 = 1, then it will produce a expr  NULL = 1 for the parquet with c1,c3 column.
After I remove the binary type coercion in the physical phase, the binary physical expr of `NULL = INT(1)` can't be created, because we don't need to support type coercion in creation physical expr.

Reply:
We need to make sure the data type of the NULL value.

Reply:
we need to get the data type from the table schema.